Subject: [PATCH] Fix ANR due to long wait for synchronization lock

Use Case: Repeated BT ON/OFF

Failure: ANR occurs due to UI wait for long time waiting
to acquire thread lock.

Steps: Repeated BT ON/OFF

Root Cause: The synchronized function that updates state
also read paired devices as an additional operation. When
the number of devices is cached list is large, the block
time for other threads waiting for same lock tends to
increase causing ANR.

Fix: Limited the synchronized block to the part where
actual update of local state takes place.

diff --git a/packages/SettingsLib/src/com/android/settingslib/bluetooth/LocalBluetoothAdapter.java b/packages/SettingsLib/src/com/android/settingslib/bluetooth/LocalBluetoothAdapter.java
old mode 100755
new mode 100644
index cda4e45..5f7ba58
--- a/packages/SettingsLib/src/com/android/settingslib/bluetooth/LocalBluetoothAdapter.java
+++ b/packages/SettingsLib/src/com/android/settingslib/bluetooth/LocalBluetoothAdapter.java
@@ -194,8 +194,13 @@
         return mState;
     }
 
-    synchronized void setBluetoothStateInt(int state) {
-        mState = state;
+    void setBluetoothStateInt(int state) {
+        synchronized(this) {
+            if (mState == state) {
+                return;
+            }
+            mState = state;
+        }
 
         if (state == BluetoothAdapter.STATE_ON) {
             // if mProfileManager hasn't been constructed yet, it will
